### PR TITLE
Shadow memory initialization functions

### PR DESCRIPTION
**Pull request by esteffin**
PR adding implementation for the Shadow memory initialization functions.

Unit tests will be added when the remaining components will be added.

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [ ] My PR is restricted to a single feature or bugfix.
- [ ] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

File: src/goto-symex/shadow_memory.cpp [hunk 1]
```diff
@@ -103,7 +105,38 @@ void shadow_memoryt::symex_field_static_init(
...
+  }
+
+  const symbolt &symbol = ns.lookup(identifier);
+
+  if(
+    (id2string(symbol.name).find("::return_value") == std::string::npos &&
+     symbol.is_auxiliary) ||
+    !symbol.is_static_lifetime)
+    return;
+  if(id2string(symbol.name).find("__cs_") != std::string::npos)
```
Comment (peterschrammel) [inline src/goto-symex/shadow_memory.cpp:143]: This case needs to be handled by some future filter option. I'd remove it for now.

File: src/goto-symex/shadow_memory.cpp [hunk 2]
```diff
@@ -170,7 +170,43 @@ void shadow_memoryt::symex_field_local_init(
...
+  if(
+    symbol_name.find("__builtin_alloca::") != std::string::npos &&
+    (symbol_name.find("alloca_size") != std::string::npos ||
+     symbol_name.find("record_malloc") != std::string::npos ||
+     symbol_name.find("record_alloca") != std::string::npos ||
+     symbol_name.find("res") != std::string::npos))
+  {
+    return;
+  }
+  if(symbol_name.find("__cs_") != std::string::npos)
```
Comment (peterschrammel) [inline src/goto-symex/shadow_memory.cpp:209]: Same here (filter option).

File: src/goto-symex/shadow_memory.cpp [hunk 3]
```diff
@@ -170,7 +170,43 @@ void shadow_memoryt::symex_field_local_init(
...
+  const symbolt &symbol =
+    ns.lookup(to_symbol_expr(expr.get_original_expr()).get_identifier());
+
+  const std::string symbol_name = id2string(symbol.name);
+  if(
+    symbol.is_auxiliary &&
+    symbol_name.find("::return_value") == std::string::npos)
+    return;
+  if(
+    symbol_name.find("malloc::") != std::string::npos &&
```
Comment (peterschrammel) [inline src/goto-symex/shadow_memory.cpp:192]: These filters should be made more precise.

File: src/goto-symex/shadow_memory.cpp (hunk 1 above)
Comment (esteffin) [reply src/goto-symex/shadow_memory.cpp:143]: We added a `TODO` comment saying that the filtering will be done when the last functionalities are added, so we will have more confidence that we are not adding regressions.

File: src/goto-symex/shadow_memory.cpp (hunk 2 above)
Comment (esteffin) [reply src/goto-symex/shadow_memory.cpp:209]: As above

File: src/goto-symex/shadow_memory.cpp (hunk 3 above)
Comment (esteffin) [reply src/goto-symex/shadow_memory.cpp:192]: As above

**Review by peterschrammel — APPROVED** [review]

**Pull request merged**